File: frontend/js/views/real_raise.js
// Real raise — whether a pay raise beats inflation, via /calc/real-raise. Live.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 3 }) + '%';

export async function renderRealRaise(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.realraise.h1.title">// REAL RAISE</span></h1>
        <p class="muted small" data-i18n="view.realraise.hint.intro">
            Whether a pay raise actually beats inflation. A nominal raise only buys more if it
            outpaces inflation — a raise equal to inflation is a wash, and below it is a real pay
            cut. Shows the inflation-adjusted break-even salary and the real (purchasing-power)
            change. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.realraise.h2.inputs">The offer</h2>
            <form id="realraise-form" class="inline-form">
                <label><span data-i18n="view.realraise.label.salary">Current salary ($)</span>
                    <input type="number" step="0.01" min="0" name="current_salary_usd" value="80000" required></label>
                <label><span data-i18n="view.realraise.label.raise">Raise (%)</span>
                    <input type="number" step="0.1" name="raise_pct" value="5" required></label>
                <label><span data-i18n="view.realraise.label.inflation">Inflation (%)</span>
                    <input type="number" step="0.1" name="inflation_pct" value="3" required></label>
            </form>
        </div>
        <div id="realraise-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#realraise-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            current_salary_usd: Number(fd.get('current_salary_usd')) || 0,
            raise_pct: Number(fd.get('raise_pct')) || 0,
            inflation_pct: Number(fd.get('inflation_pct')) || 0,
        };
        try {
            const r = await api.calcRealRaise(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.realraise.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#realraise-result');
    const cls = r.is_real_raise ? 'pos' : 'neg';
    const verdictKey = r.is_real_raise ? 'view.realraise.verdict.real' : 'view.realraise.verdict.cut';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.realraise.h2.result">After inflation</h2>
            <div class="cards">
                <div class="card ${cls}"><div class="label" data-i18n="view.realraise.card.real">Real raise</div>
                    <div class="value ${cls}">${pct(r.real_raise_pct)}</div></div>
                <div class="card ${cls}"><div class="label" data-i18n="view.realraise.card.change">Real change</div>
                    <div class="value ${cls}">${money(r.real_change_usd)}</div></div>
                <div class="card ${cls}"><div class="label" data-i18n="view.realraise.card.verdict">Verdict</div>
                    <div class="value ${cls}" data-i18n="${verdictKey}">—</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.realraise.row.new">New salary</td><td>${money(r.new_salary_usd)}</td></tr>
                    <tr><td data-i18n="view.realraise.row.nominal">Nominal change</td><td>${money(r.nominal_change_usd)}</td></tr>
                    <tr><td data-i18n="view.realraise.row.breakeven">Break-even salary (keeps pace)</td><td>${money(r.inflation_adjusted_salary_usd)}</td></tr>
                    <tr><td data-i18n="view.realraise.row.keeppace">Raise to keep pace</td><td>${pct(r.raise_to_keep_pace_pct)}</td></tr>
                    <tr class="emph ${cls}"><td data-i18n="view.realraise.row.real">Real raise</td><td>${pct(r.real_raise_pct)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/real_raise.rs
//! Real raise — whether a pay raise actually beats inflation, and the raise
//! needed just to keep pace.
//!
//! A nominal raise only buys more if it outpaces inflation. The real change
//! uses the Fisher relation:
//!
//! ```text
//! inflation-adjusted salary = current × (1 + inflation)   (break-even pay)
//! real raise = (1 + raise) / (1 + inflation) − 1
//! ```
//!
//! A raise equal to inflation is a wash; below it, a real pay cut.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RealRaiseInput {
    pub current_salary_usd: f64,
    pub inflation_pct: f64,
    pub raise_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct RealRaiseResult {
    /// Raise needed just to keep pace with inflation (= inflation rate).
    pub raise_to_keep_pace_pct: f64,
    /// Salary that would merely maintain purchasing power.
    pub inflation_adjusted_salary_usd: f64,
    pub new_salary_usd: f64,
    /// New − current (nominal dollars).
    pub nominal_change_usd: f64,
    /// Inflation-adjusted change, percent (the real raise).
    pub real_raise_pct: f64,
    /// New salary − inflation-adjusted salary (purchasing-power gain/loss).
    pub real_change_usd: f64,
    /// Whether the raise beats inflation.
    pub is_real_raise: bool,
}

pub fn analyze(input: &RealRaiseInput) -> RealRaiseResult {
    let infl = input.inflation_pct / 100.0;
    let raise = input.raise_pct / 100.0;

    let inflation_adjusted = input.current_salary_usd * (1.0 + infl);
    let new_salary = input.current_salary_usd * (1.0 + raise);
    let real_raise = if (1.0 + infl) != 0.0 {
        ((1.0 + raise) / (1.0 + infl) - 1.0) * 100.0
    } else {
        0.0
    };

    RealRaiseResult {
        raise_to_keep_pace_pct: input.inflation_pct,
        inflation_adjusted_salary_usd: inflation_adjusted,
        new_salary_usd: new_salary,
        nominal_change_usd: new_salary - input.current_salary_usd,
        real_raise_pct: real_raise,
        real_change_usd: new_salary - inflation_adjusted,
        is_real_raise: input.raise_pct > input.inflation_pct,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-4
    }

    fn run(salary: f64, infl: f64, raise: f64) -> RealRaiseResult {
        analyze(&RealRaiseInput {
            current_salary_usd: salary,
            inflation_pct: infl,
            raise_pct: raise,
        })
    }

    #[test]
    fn inflation_adjusted_salary() {
        // 80,000 × 1.03 = 82,400.
        assert!(close(run(80_000.0, 3.0, 5.0).inflation_adjusted_salary_usd, 82_400.0));
    }

    #[test]
    fn new_salary_and_nominal_change() {
        let r = run(80_000.0, 3.0, 5.0);
        assert!(close(r.new_salary_usd, 84_000.0));
        assert!(close(r.nominal_change_usd, 4_000.0));
    }

    #[test]
    fn real_raise_pct() {
        // (1.05/1.03 − 1) × 100 = 1.941748%.
        assert!(close(run(80_000.0, 3.0, 5.0).real_raise_pct, 1.941748));
    }

    #[test]
    fn real_change_usd() {
        // 84,000 − 82,400 = 1,600.
        assert!(close(run(80_000.0, 3.0, 5.0).real_change_usd, 1_600.0));
    }

    #[test]
    fn beats_inflation() {
        assert!(run(80_000.0, 3.0, 5.0).is_real_raise);
    }

    #[test]
    fn real_cut_below_inflation() {
        let r = run(80_000.0, 3.0, 2.0);
        assert!(!r.is_real_raise);
        assert!(r.real_raise_pct < 0.0);
        assert!(r.real_change_usd < 0.0);
    }

    #[test]
    fn raise_equals_inflation_is_a_wash() {
        let r = run(80_000.0, 3.0, 3.0);
        assert!(close(r.real_raise_pct, 0.0));
        assert!(close(r.real_change_usd, 0.0));
        assert!(!r.is_real_raise);
    }

    #[test]
    fn keep_pace_equals_inflation() {
        assert!(close(run(80_000.0, 3.0, 5.0).raise_to_keep_pace_pct, 3.0));
    }
}
